core/storage/suggestion/gae_models.py

```diff
@@ -0,0 +1,267 @@
+]
+
+# Defines what is the minimum role required to review suggestions
+# of a particular type.
+SUGGESTION_MINIMUM_ROLE_FOR_REVIEW = {
+    SUGGESTION_EDIT_STATE_CONTENT: feconf.ROLE_ID_EXPLORATION_EDITOR
+}
+
+# Constants defining various contribution types.
+CONTRIBUTION_TYPE_CONTENT = 'content'
+CONTRIBUTION_TYPE_TRANSLATION = 'translation'
+
+CONTRIBUTION_TYPE_CHOICES = [
+    CONTRIBUTION_TYPE_CONTENT,
+    CONTRIBUTION_TYPE_TRANSLATION
+]
+
+#The delimiter to be used in the suggestion ID.
+SUGGESTION_ID_DELIMITER = '.'
+
+
+class SuggestionParams(ndb.Model):
+    """Model to store all the suggestion params for a suggestion."""
+
+    # The contribution type.
+    contribution_type = ndb.StringProperty(
+        required=True, choices=CONTRIBUTION_TYPE_CHOICES)
+    # The contribution domain. This parameter along with the above contribution
+    # type will be used to determine the category to 'score' the suggester.
+    contribution_domain = ndb.StringProperty(required=True)
+    # The ID of the entity being suggested to.
+    entity_id = ndb.StringProperty(required=True)
+    # The version number of the entity being suggested to.
+    entity_version = ndb.IntegerProperty(required=True)
+
+
+class SuggestionModel(base_models.BaseModel):
+    """Model to store suggestions made by Oppia users."""
+
+    # The type of suggestion.
+    suggestion_type = ndb.StringProperty(
+        required=True, indexed=True, choices=SUGGESTION_TYPE_CHOICES)
+    # The type of entity which the suggestion is linked to.
+    entity_type = ndb.StringProperty(
+        required=True, indexed=True, choices=ENTITY_TYPE_CHOICES)
+    # Status of the suggestion.
+    status = ndb.StringProperty(
+        required=True, indexed=True, choices=STATUS_CHOICES)
+    # Additional parameters to be stored.
+    suggestion_params = ndb.StructuredProperty(SuggestionParams)
+    # The author of the suggestion.
+    author_id = ndb.StringProperty(required=True, indexed=True)
+    # The reviewer who accepted the suggestion.
+    reviewer_id = ndb.StringProperty(required=False, indexed=True)
+    # The thread linked to this suggestion.
+    thread_id = ndb.StringProperty(required=True, indexed=True)
+    # The reviewer assigned to review the suggestion.
+    assigned_reviewer_id = ndb.StringProperty(required=True, indexed=True)
+    # The suggestion payload. Contains the data of the suggestion. For "edit"
+    # suggestions, this contains the change_list. For "add" suggestions, this
+    # contains the data necessary to add the new entity.
+    payload = ndb.JsonProperty(required=True, indexed=False)
+
+    @classmethod
+    def get_instance_id(
+            cls, entity_type, thread_id, entity_id):
+        """Concatenates various parameters and gives the ID of the suggestion
+        model.
+
+        Args:
+            entity_type: str. The type of entity being edited/added.
+            thread_id: str. The ID of the feedback thread linked to the
```

Comment:
See my comment in the design doc. Feedback thread IDs contain exploration ID and dots, right? Isn't this going to cause issues with the delimiter used here, especially since you're sandwiching thread ID between a couple of other things? It creates ambiguity in parsing.

Also it's a bit weird that entity type and ID aren't adjacent.

Reply:
So I plan to use only the instance ID part of the feedback thread ID as the thread_id here. That part of the ID is just a randomly generated string. So it should be fine. I think I got confused about the thread IDs after looking at generate_new_thread_id function in feedback/gae_models.py which returns just the instance ID. I thought that was the thread ID as such. The output of that function is what I plan to pass into this function.

Reply:
(NB: We talked about making the feedback thread ID and suggestion ID the same. I think that might be the most robust solution.)

Reply:
Done from the suggestions perspective, will migrate feedback threads later in this milestone.